### DISCUSSION TITLE
Can we decouple API routes from file based routing

### DISCUSSION THREAD

**Discussion by francisashley** [nuxt/nuxt#24968] [Ideas]
The current approach is verbose and frustrating to use for advanced projects. Ideally we have the option to register all API routes in a single roots file and pass in handler functions ourselves.

Currently the closest thing we can do to is register custom paths to files to handle the response (file based routing):

```ts
import { defineNuxtModule } from '@nuxt/kit'

const endpoints = [{ method: 'get', route: '/api/some/route', handler: '@/some/api/route-handler.get.ts' }]

export default defineNuxtModule({
  hooks: {
    'nitro:config'(nitroConfig) {
      endpoints.forEach((endpoint) => {
        nitroConfig.handlers?.push(endpoint)
      })
    },
  },
})
```
If the handler field can be changed to also support a function, this would be so much cleaner to work with:

```ts
import { defineNuxtModule } from "@nuxt/kit";

const endpoints = [
  {
    method: "get",
    route: "/api/some/route",
    handler: defineEventHandler((event) => console.log("handle request")),
  },
];

export default defineNuxtModule({
  hooks: {
    "nitro:config"(nitroConfig) {
      endpoints.forEach((endpoint) => {
        nitroConfig.handlers?.push(endpoint);
      });
    },
  },
});

```

**Comment by francisashley** [comment]
I have also posted about this in the nitro repo https://github.com/unjs/nitro/discussions/2042

**Comment by masyoudi** [comment]
do you  need define routing like this?

```ts
import { createRouter, useBase } from 'h3';
import test from '~~/server/controllers/test.ts';

const router = createRouter();

router.get('/test', test.get);

export default useBase('/api', router.handler);
```

[see this example](https://github.com/masyoudi/nuxt3sse/blob/master/server/api/%5B...%5D.ts)

**Reply by francisashley** [reply]
OMG this is PERFECT! This was my one frustration with Nuxt, now it's perfect. Thank you! 🙌
